extensions/multiline.md

```diff
@@ -0,0 +1,241 @@
+
+Clients MUST ignore every token with a key that they don't understand.
+
+The only defined capability key so far is:
+
+* `max-bytes` - This defines the maximum allowed total byte length of multiline batched content *REQUIRED*
+* `max-lines` - This defines the maximum allowed number of lines in a multiline batch content *OPTIONAL*
+
+### Tags
+
+This specification adds the `draft/multiline-concat` message tag, which has no value and is used to send message lines that extend beyond the usual byte length limit. Its usage is described below.
+
+### Batch types
+
+This specification adds the `draft/multiline` batch type, and introduces client initiated batches. These use the same syntax as server initiated batches.
+
+In addition to the base batch parameters (reference-tag and type) a multiline batch has one additional parameter, the target recipient.
+
+Multiline batches MUST contain one or more PRIVMSG lines whose target MUST match the batch target.
+
+When receiving a well-formed mulitiline message batch, implementations MUST collect each PRIVMSG message content and wait until the full batch has been received before processing the message. Processing in this context refers to:
+
+* Servers: delivering the batch to the intended recipients
+* Clients: displaying the batched message to the user
+
+Messages in a multiline batch MUST be concatenated with a line break unless the `draft/multiline-concat` message tag is sent, in which case the message is directly concatenated with the previous message without any separation.
+
+Clients MUST NOT send messages other than PRIVMSG while a multiline batch is open.
+
+### Fallback
+
+When delivering multiline batches to clients that have not negotiated the multiline capability, servers MUST deliver the component messages without using a multiline BATCH.
+
+Any tags that would have been added to the batch, e.g. message IDs, account tags etc MUST be included on the first message line. Tags MAY also be included on subsequent lines where it makes sense to do so.
+
+### Errors
+
+Servers MUST use `FAIL` messages from the [standard replies][] framework to notify clients of errors with multiline batches. The command is `BATCH` and the following codes are defined:
+
+* `MULTILINE_MAX_BYTES <limit>`: `max-bytes` limit exceeded
+* `MULTILINE_MAX_LINES <limit>`: `max-lines` limit exceeded
+* `MULTILINE_INVALID_TARGET <batch-target> <provided-target>`: mismatched PRIVMSG target within a batch
+* `MULTILINE_INVALID`: any other error
+
+## Interaction with other specifications
+
+### Message tags ([spec][message tags])
+
+Clients MUST only send non-`batch` tags on the opening BATCH command when sending multiline batches.
+
+### Message ids ([spec][message ids])
+
+Servers MUST only include a message ID on the first message of a batch when sending a fallback to non-supporting clients.
```

Comment:
Why the first one? If the message ID is used eg for a reply or a reaction emoji, it visually makes more sense to have it attached to the last message.

Reply:
If I understand correctly, this is saying that (some of) the fallback `PRIVMSG` should be sent without any `msgid` tag at all? This seems to upend some expectations (e.g., msgids being available for deduplication purposes).

Reply:
For comparison, [oragono.io/maxline-2](https://oragono.io/maxline-2) generates n+1 distinct message IDs for the original message and all n lines of the split message. This isn't perfect, but it maintains the expectation that `PRIVMSG` should have an ID.

The msgid spec says: "If this tag is being used, it SHOULD be attached to all PRIVMSG and NOTICE events."

Reply:
msgids can't appear more than once. Generating unique n+1 msgids and only sending them on the fallback creates undesirable situations where people might reply to a fallback message and the context would then be missing for people who fully support maxline (i.e. discard fallback message tags)

If you're concerned about missing msgids, the solution would be to properly support the maxline batch spec.

Reply:
I don't think that's true. It'd be easy to generate unique fallback msgids that the server can recognise as such and translate back or refuse replies for or something.

Reply:
@jwheare, just to be clear, when you say "maxline" you're referring to the present multiline spec, not to oragono.io/maxline-2?

Some degradation of experience is inevitable under any proposal to support PRIVMSG over 512 bytes. I'm much less concerned about the case you're describing than I am about sending PRIVMSG without IDs, which seems to complicate the handling of chat history at a more fundamental level.

I'm not 100% sold on @edk0's suggestion because it violates an expectation I have that servers shouldn't need to care about client-only tags, but I could be convinced.

Hypothetically, we could define a new tag like `draft/fmsgid` ("fallback msgid") for transmitting the fallback msgids to clients that support the multiline spec.

Reply:
I'm going to care about client-only tags anyway, so it's not a bother for me. You don't have to implement my suggestion; I'm just saying a solution is possible.